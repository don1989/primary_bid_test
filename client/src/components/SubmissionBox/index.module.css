.root {
	display: flex;
	flex-direction: column;
	height: 400px;
	width: 400px;
	box-shadow: 2px 2px;
	border: solid 2px blueviolet;
	border-radius: 20px;
}
