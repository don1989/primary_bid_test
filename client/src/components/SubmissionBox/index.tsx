import React, { useState } from 'react';
import styles from './index.module.css';
import { postUrl } from '../../requests/urls';
import { AxiosResponse } from 'axios';
import { IPostResponseBody } from '../../types';

interface Props {
	onSubmitSuccess: () => void;
}

const SubmissionBox = ({ onSubmitSuccess }: Props) => {
	const [url, setUrl] = useState<string>('');
	const [didSubmit, setDidSubmit] = useState<boolean>(false);
	const [shortUrl, setShortUrl] = useState<string>('');
	const [didCopy, setDidCopy] = useState<boolean>(false);
	const [timeoutId, setTimeoutId] = useState<NodeJS.Timeout>();

	const handleSubmit = () => {
		setDidSubmit(true);

		postUrl(url)
			.then((res: AxiosResponse<IPostResponseBody>) => {
				setShortUrl(res.data.shortUrl);
				onSubmitSuccess();
			})
			.catch((err) => {
				// TODO: display err
			})
			.finally(() => {
				setDidSubmit(false);
			});
	};

	const handleChangeText = (event: React.ChangeEvent<HTMLInputElement>) => {
		setUrl(event.target.value);
	};

	const handleCopyToClipboard = () => {
		navigator.clipboard.writeText(shortUrl);
		setDidCopy(true);
		if (timeoutId) {
			clearTimeout(timeoutId);
		}
		const id = setTimeout(() => {
			setDidCopy(false);
			setTimeoutId(undefined);
		}, 4000);
		setTimeoutId(id);
	};

	const handleReset = () => {
		setUrl('');
		setShortUrl('');
	};

	return (
		<div className={styles.root}>
			<div>
				<h2>Enter long URL</h2>
				<input
					type="text"
					value={url}
					onChange={handleChangeText}
					readOnly={!!shortUrl}
				/>
				<button
					disabled={!url || didSubmit || !!shortUrl}
					onClick={handleSubmit}
				>
					Submit
				</button>

				{shortUrl && (
					<div>
						<h2>Short URL</h2>
						<input type="text" value={shortUrl} readOnly={true} />
						<button onClick={handleCopyToClipboard}>Copy</button>
						<button onClick={handleReset}>Shorten Another</button>
					</div>
				)}

				{didCopy && <div>Copied to clipboard</div>}
			</div>
		</div>
	);
};

export default SubmissionBox;
