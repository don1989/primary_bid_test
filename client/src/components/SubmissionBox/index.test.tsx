import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import SubmissionBox from './';

describe('SubmissionBox', () => {
	test('renders a list item', () => {
		const onSubmitSuccess = jest.fn();
		const { asFragment, getByText } = render(
			<SubmissionBox onSubmitSuccess={onSubmitSuccess} />
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
