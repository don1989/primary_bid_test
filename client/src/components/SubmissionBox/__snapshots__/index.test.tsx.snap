// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmissionBox renders a list item 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div>
      <h2>
        Enter long URL
      </h2>
      <input
        type="text"
        value=""
      />
      <button
        disabled=""
      >
        Submit
      </button>
    </div>
  </div>
</DocumentFragment>
`;
